optimisation/design/Third_Party_API_Design.md

```diff
@@ -0,0 +1,595 @@
+(A workaround for this for the Python user interface using a compiled Horace library exists,
+and is described [here](../../python_interface/design/01_pace_python_high_level_discussion.md#interpreter-calls)).
+This means that the second type of function signatures described in the [optimisation design document](Model_Optimisation_Design.md#optimisation-function-object), 
+where Matlab objects (in particular an `sqw` object) are passed as input to the user defined function,
+will not be possible with non-Matlab functions, unless `OptFunction` unwraps this object into a (several) POD array(s).
+
+The next difficulty is to infer the required input arguments for user defined model functions.
+Several methods for this springs to mind:
+
+1. Using properties of `OptFunction` to indicate the input that the user defined model function takes as described [here](Model_Optimisation_Design.md#optimisation-function-object).
+2. By Horace interrogating the function as to its input requirements by calling it with a single special argument.
+3. Embedded in the function name - for example functions with suffix `_sqw` are assumed to take crystal coordinates ![\mathbf{Q}] as inputs.
+4. Infer input arguments using language specific constructs (e.g. the `inspect` module in Python, C++ templates).
+
+Which of these methods is best depends on the language of the user defined model function and is a topic for discussion.
+The first method would be easiest to implement for all languages.
+The second method will likely only work for Matlab and Python,
+since variable number of arguments is difficult to implement in compiled languages.
+Method 3 (inference from function name) will probably work for all languages,
+but may prove difficult to implement for the compiled languages.
+Method 4 would only work for Python and C++.
+
+As discussed in the section on [user models written in compiled languages](#compiled) below,
+Matlab has two ways to call functions in a compiled library,
+the simpler of which uses a C interface (not C++).
+*As noted there, we propose to use this C-style approach, in order to present a simple and uniform API for all compiled languages.*
+However, this means that methods 2 and 4 for inferring input arguments are not feasible,
+and method 3 would be difficult to implement as discussed in the [section on the C API](#c_functions).
+
+
+## <a name="python"></a> Python
+
+There are two use cases for user defined model functions written in Python:
+
+1. The user is running Matlab (has a license) and wants to call a model function (e.g. Euphonic) written in Python.
+2. The user is running Horace in Python using the compiled library and wants to call their own Python model function.
+
+Taking the first case, it is currently possible to do this in Matlab
+(as exemplified by the [`meuphonic` project](https://github.com/pace-neutrons/Meuphonic)).
```

Comment:
`meuphonic` has been renamed [`horace-euphonic-interface`](https://github.com/pace-neutrons/horace-euphonic-interface), however the old link does and will forever continue to work.